### ISSUE TITLE
Assert.Equivalent detects circular reference in Uri

### ISSUE THREAD

**Issue by chris468** [xunit/xunit#2939]
If `Assert.Equivalent` is called on values that contain `Uri`s, it asserts that there is a circular reference.

Reproducing project [here](https://gist.github.com/chris468/9bf4c17dc9313b79df5965827fe9575e). `dotnet new xunit` w/ sdk version 8.0.300 and assert equivalence on Uri. (Originally discovered in a more complex hierarchy).

```csharp
[Fact]
public void Test1()
{
    Uri expected = new Uri("http://example.com");
    Uri actual = new Uri("http://example.com");
    Assert.Equivalent(expected, actual, strict: true);
}
```

```
[xUnit.net 00:00:00.14]     equivalent_uri.UnitTest1.Test1 [FAIL]
  Failed equivalent_uri.UnitTest1.Test1 [5 ms]
  Error Message:
   Assert.Equivalent() Failure: Circular reference found in 'expected.AbsoluteUri'
  Stack Trace:
     at equivalent_uri.UnitTest1.Test1() in C:\tmp\equivalent-uri\UnitTest1.cs:line 10
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)

Failed!  - Failed:     1, Passed:     0, Skipped:     0, Total:     1, Duration: < 1 ms - equivalent-uri.dll (net8.0)
```

I believe the issue is that the `HashSet` [used to detect circular references](https://github.com/xunit/assert.xunit/blob/a062fd1fee36945168c3e419c63980c20ffd7717/Sdk/AssertHelper.cs#L307) uses equality comparison. Uri is comparing equal to the equivalent string, and for absolute Uris the `AbsoluteUri` property compares equal to the `Uri` itself.

**Comment by bradwilson** [comment]
No, the reality is that `Assert.Equivalent` doesn't work for objects which are internally circular in nature, and the way Uri is defined here makes it appear to be circular in nature.

I'll take a look and see if there is an obvious thing that can be done here, but I'm skeptical.

**Comment by bradwilson** [comment]
Ignore my previous post. 😂 Now fully understanding the problem (🤦🏼‍♂️), the fix is easy. Switching the hash set to reference comparison instead of an equality comparison does indeed fix the problem.

**Comment by bradwilson** [comment]
Available in v2 `2.8.2-pre.9`
Available in v3 `0.1.1-pre.441`

https://xunit.net/docs/using-ci-builds
